src/RegionCosts.cpp

```diff
@@ -0,0 +1,706 @@
+#include "RegionCosts.h"
+
+namespace Halide {
+namespace Internal {
+
+using std::string;
+using std::map;
+using std::pair;
+using std::set;
+using std::vector;
+using std::make_pair;
+using std::deque;
+
+
+bool has_suffix(const std::string &str, const std::string &suffix) {
+    return str.size() >= suffix.size() &&
+            str.compare(str.size() - suffix.size(),
+                        suffix.size(), suffix) == 0;
+}
+
+struct ExprCost : public IRVisitor {
+    int64_t ops;
+    int64_t byte_loads;
+    map<string, int64_t> detailed_byte_loads;
+
+    ExprCost() {
+        ops = 0; byte_loads = 0;
+    }
+
+    using IRVisitor::visit;
+
+    void visit(const IntImm *) {}
+    void visit(const UIntImm *) {}
+    void visit(const FloatImm *) {}
+    void visit(const StringImm *) {}
+    void visit(const Cast * op) {
+        op->value.accept(this);
+        ops+=1;
+    }
+    void visit(const Variable *) {}
+
+    template<typename T>
+        void visit_binary_operator(const T *op, int cost) {
+            op->a.accept(this);
+            op->b.accept(this);
+            ops += cost;
+        }
+
+    // TODO: Figure out the right costs
+    void visit(const Add *op) {visit_binary_operator(op, 1);}
+    void visit(const Sub *op) {visit_binary_operator(op, 1);}
+    void visit(const Mul *op) {visit_binary_operator(op, 1);}
+    void visit(const Div *op) {visit_binary_operator(op, 1);}
+    void visit(const Mod *op) {visit_binary_operator(op, 1);}
+    void visit(const Min *op) {visit_binary_operator(op, 1);}
+    void visit(const Max *op) {visit_binary_operator(op, 1);}
+    void visit(const EQ *op) {visit_binary_operator(op, 1);}
+    void visit(const NE *op) {visit_binary_operator(op, 1);}
+    void visit(const LT *op) {visit_binary_operator(op, 1);}
+    void visit(const LE *op) {visit_binary_operator(op, 1);}
+    void visit(const GT *op) {visit_binary_operator(op, 1);}
+    void visit(const GE *op) {visit_binary_operator(op, 1);}
+    void visit(const And *op) {visit_binary_operator(op, 1);}
+    void visit(const Or *op) {visit_binary_operator(op, 1);}
+
+    void visit(const Not *op) {
+        op->a.accept(this);
+        ops+=1;
+    }
+
+    void visit(const Select *op) {
+        op->condition.accept(this);
+        op->true_value.accept(this);
+        op->false_value.accept(this);
+        ops+=1;
+    }
+
+    void visit(const Call * call) {
+        if (call->call_type == Call::Halide ||
+                call->call_type == Call::Image) {
```

Comment:
This expression should probably be on a single line: Halide doesn't require 80-column line lengths, and this would probably read better as a single line. (Ditto several other places here.)